File: Top_100_Liked_Questions/380._Insert_Delete_GetRandom_O(1)/solution.js
/**
 * Initialize your data structure here.
 */
var RandomizedSet = function() {
    this.hashmap = {};
    this.valArray = [];
};

/**
 * Inserts a value to the set. Returns true if the set did not already contain the specified element. 
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.insert = function(val) {
    if (!this.hashmap.hasOwnProperty(val)) {
        this.hashmap[val] = this.valArray.length;
        this.valArray.push(val);
        return true;
    }
    
    return false;
};

/**
 * Removes a value from the set. Returns true if the set contained the specified element. 
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.remove = function(val) {
    if (this.hashmap.hasOwnProperty(val)) {
        let idx = this.hashmap[val];
        let tailVal = this.valArray.pop();
        if (idx != this.valArray.length) {
            this.valArray[idx] = tailVal;
            this.hashmap[tailVal] = idx;
        }
        delete this.hashmap[val];
        return true;
    }
    
    return false;
};

/**
 * Get a random element from the set.
 * @return {number}
 */
RandomizedSet.prototype.getRandom = function() {
    return this.valArray[Math.floor(Math.random() * this.valArray.length)];
};

/** 
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = new RandomizedSet()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */

File: Top_100_Liked_Questions/380._Insert_Delete_GetRandom_O(1)/solution.py
class RandomizedSet:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.valLst = []
        self.hashIdx = {}

    def insert(self, val: int) -> bool:
        """
        Inserts a value to the set. Returns true if the set did not already contain the specified element.
        """
        if val not in self.hashIdx:
            self.hashIdx[val] = len(self.valLst)
            self.valLst.append(val)
            return True
        
        return False

    def remove(self, val: int) -> bool:
        """
        Removes a value from the set. Returns true if the set contained the specified element.
        """
        if val in self.hashIdx:
            idx = self.hashIdx[val]
            tailVal = self.valLst.pop()
            if idx != len(self.valLst):
                self.valLst[idx] = tailVal
                self.hashIdx[tailVal] = idx
            del self.hashIdx[val]
            return True
        
        return False

    def getRandom(self) -> int:
        """
        Get a random element from the set.
        """
        return random.choice(self.valLst)


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()